libs/beacon/benchmark/entropy_gen_bench.cpp

```diff
@@ -125,134 +115,113 @@ struct BeaconSelfContained
 };
 
 void EntropyGen(benchmark::State &state)
 {
-  char const *LOGGING_NAME = "EntropyGen";
-
   SetGlobalLogLevel(LogLevel::ERROR);
 
-  uint16_t unique_port  = 8000;
-  uint16_t test_attempt = 0;
+  uint16_t unique_port    = 8000;
+  uint16_t test_attempt   = 0;
+  uint32_t entropy_rounds = 10;
 
   std::vector<std::unique_ptr<BeaconSelfContained>> nodes;
+  BeaconService::CabinetMemberList                  cabinet;
+  auto nodes_in_test = static_cast<uint64_t>(state.range(0));
 
-  for (auto _ : state)
+  // Setup here - not included in test timing
   {
-    BeaconServiceInsertable::CabinetMemberList cabinet;
-
-    uint64_t nodes_in_test = uint64_t(state.range(0));
+    nodes.clear();
+    nodes.resize(nodes_in_test);
 
-    FETCH_LOG_INFO(LOGGING_NAME, "===============================");
-    FETCH_LOG_INFO(LOGGING_NAME, "Starting test: ", nodes_in_test);
+    for (uint16_t i = 0; i < nodes_in_test; ++i)
+    {
+      auto &node = nodes[i];
+      if (!node)
+      {
+        node = std::make_unique<BeaconSelfContained>(unique_port + i, i);
+        node->Start();
+      }
+      cabinet.insert(node->muddle_certificate->identity().identifier());
+    }
 
-    // Setup here - not included in test timing
+    for (uint16_t i = 0; i < nodes_in_test; ++i)
     {
-      state.PauseTiming();
-      nodes.clear();
-      test_attempt++;
+      auto &node = nodes[i];
+      for (uint16_t j = 0; j < i; j++)
+      {
+        node->muddle->ConnectTo(nodes[j]->GetMuddleAddress(), nodes[j]->GetHint());
+      }
+    }
 
-      nodes.resize(nodes_in_test);
+    // Wait until fully connected
+    while (true)
+    {
+      bool connected = true;
 
-      for (uint16_t i = 0; i < nodes_in_test; ++i)
+      for (auto const &node : nodes)
       {
-        auto &node = nodes[i];
-        if (!node)
-        {
-          node = std::make_unique<BeaconSelfContained>(unique_port + i, i);
-          node->Start();
-        }
-        cabinet.insert(node->muddle_certificate->identity().identifier());
-
-        for (uint16_t j = 0; j < i; j++)
+        if (node->muddle->GetNumDirectlyConnectedPeers() != nodes_in_test - 1)
         {
-          node->muddle->ConnectTo(nodes[j]->GetMuddleAddress(), nodes[j]->GetHint());
+          connected = false;
         }
       }
 
-      for (auto const &member : cabinet)
+      if (connected)
       {
-        FETCH_LOG_INFO(LOGGING_NAME, "Cabinet member: ", member.ToBase64());
+        break;
       }
+    }
+  }
 
-      for (auto const &node : nodes)
-      {
-        node->ResetCabinet(cabinet);
-      }
+  for (auto _ : state)
+  {
+    state.PauseTiming();
+    // Create previous entropy
+    BlockEntropy prev_entropy;
+    prev_entropy.group_signature = "Hello";
+    TrustedDealer dealer{cabinet, static_cast<uint32_t>(cabinet.size() / 2 + 1)};
+    state.ResumeTiming();
+    for (auto const &node : nodes)
+    {
+      node->ResetCabinet(cabinet, test_attempt * entropy_rounds,
```

Comment:
I might have included this reset + filling pending_nodes inside the test pausing. There is also an efficient way to fill a vector incrementally if you like:

```
    std::vector<int> thing(10);
    std::iota(thing.begin(), thing.end(), 0);
```

Reply:
I moved this outside the resume timing because once some members start sending signatures it is possible for a member to have have already generated the entropy before their cabinet is reset from receiving signature shares. I will move the filling outside of the for loop.